tokio/src/sync/rwlock.rs

```diff
@@ -0,0 +1,240 @@
+/// system's implementation, and this type does not guarantee that any
+/// particular policy will be used.
+///
+/// The type parameter `T` represents the data that this lock protects. It is
+/// required that `T` satisfies [`Send`] to be shared across threads. The RAII guards
+/// returned from the locking methods implement [`Deref`](https://doc.rust-lang.org/std/ops/trait.Deref.html)
+/// (and [`DerefMut`](https://doc.rust-lang.org/std/ops/trait.DerefMut.html)
+/// for the `write` methods) to allow access to the content of the lock.
+///
+/// # Examples
+///
+/// ```
+/// use tokio::sync::RwLock;
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let lock = RwLock::new(5);
+///
+/// // many reader locks can be held at once
+///     {
+///         let r1 = lock.read().await;
+///         let r2 = lock.read().await;
+///         assert_eq!(*r1, 5);
+///         assert_eq!(*r2, 5);
+///     } // read locks are dropped at this point
+///
+/// // only one write lock may be held, however
+///     {
+///         let mut w = lock.write().await;
+///         *w += 1;
+///         assert_eq!(*w, 6);
+///     } // write lock is dropped here
+/// }
+/// ```
+///
+/// [`Mutex`]: struct.Mutex.html
+/// [`RwLock`]: struct.RwLock.html
+/// [`RwLockReadGuard`]: struct.RwLockReadGuard.html
+/// [`RwLockWriteGuard`]: struct.RwLockWriteGuard.html
+/// [`Send`]: https://doc.rust-lang.org/std/marker/trait.Send.html
+#[derive(Debug)]
+pub struct RwLock<T> {
+    //semaphore to coordinate read and write access to T
+    s: Semaphore,
+
+    //inner data T
+    c: UnsafeCell<T>,
+
+    //mutex to coordinate writes
+    w: Mutex<()>,
+}
+
+/// RAII structure used to release the shared read access of a lock when
+/// dropped.
+///
+/// This structure is created by the [`read`] method on
+/// [`RwLock`].
+///
+/// [`read`]: struct.RwLock.html#method.read
+#[derive(Debug)]
+pub struct RwLockReadGuard<'a, T> {
+    permit: ReleasingPermit<'a, T>,
+    lock: &'a RwLock<T>,
+}
+
+/// RAII structure used to release the exclusive write access of a lock when
+/// dropped.
+///
+/// This structure is created by the [`write`] and method
+/// on [`RwLock`].
+///
+/// [`write`]: struct.RwLock.html#method.write
+/// [`RwLock`]: struct.RwLock.html
+#[derive(Debug)]
+pub struct RwLockWriteGuard<'a, T> {
+    permits: Vec<ReleasingPermit<'a, T>>,
+    lock: &'a RwLock<T>,
+}
+
+// Wrapper arround Permit that releases on Drop
+#[derive(Debug)]
+struct ReleasingPermit<'a, T>(Permit, &'a RwLock<T>);
+
+impl<'a, T> Drop for ReleasingPermit<'a, T> {
+    fn drop(&mut self) {
+        self.0.release(&self.1.s);
+    }
+}
+
+// As long as T: Send, it's fine to send and share RwLock<T> between threads.
+// If T were not Send, sending and sharing a RwLock<T> would be bad, since you can access T through
+// RwLock<T>.
+unsafe impl<T> Send for RwLock<T> where T: Send {}
+unsafe impl<T> Sync for RwLock<T> where T: Send {}
```

Comment:
I believe this bound should be:
```rust
unsafe impl<T> Sync for RwLock<T> where T: Send + Sync {}
```
(Similar to how it is in std: <https://doc.rust-lang.org/src/std/sync/rwlock.rs.html#76>)

Otherwise it would be possible to use a Tokio `RwLock` to allow concurrent access to a `!Sync` type across threads, violating its contract:
```rust
fn read_shared_non_sync() {
    let rwlock = Arc::new(RwLock::<Foo>::new(Foo(&mut ())));

    let rw = rwlock.clone();

    let h1 = std::thread::spawn(move || {
        let mut r = spawn(rw.read());
        let r = assert_ready!(r.poll());
    });
    
    let rw = rwlock.clone();

    let h2 = std::thread::spawn(move || {
        let mut r = spawn(rw.read());
        let r = assert_ready!(r.poll());
    });

    h1.join().unwrap();
    h2.join().unwrap();

    struct Foo(*mut ());

    unsafe impl Send for Foo {
    }
}
```

Reply:
yes thanks, updated it!